File: Tellma/ClientApp/src/app/data/entities/relation.ts
// tslint:disable:variable-name
// tslint:disable:max-line-length
import { EntityWithKey } from './base/entity-with-key';
import { WorkspaceService } from '../workspace.service';
import { TranslateService } from '@ngx-translate/core';
import { EntityDescriptor, NavigationPropDescriptor, NumberPropDescriptor } from './base/metadata';
import { SettingsForClient } from '../dto/settings-for-client';
import { DefinitionsForClient } from '../dto/definitions-for-client';
import { RelationUserForSave, RelationUser } from './relation-user';
import { EntityForSave } from './base/entity-for-save';
import { RelationAttachment, RelationAttachmentForSave } from './relation-attachment';

export interface RelationForSave<TRelationUser = RelationUserForSave, TAttachment = RelationAttachmentForSave> extends EntityForSave {
  // Common with Resource
  Name?: string;
  Name2?: string;
  Name3?: string;
  Code?: string;
  CurrencyId?: string;
  CenterId?: number;
  Description?: string;
  Description2?: string;
  Description3?: string;
  LocationJson?: string;
  FromDate?: string;
  ToDate?: string;
  DateOfBirth?: string;
  ContactEmail?: string;
  ContactMobile?: string;
  NormalizedContactMobile?: string;
  ContactAddress?: string;
  Date1?: string;
  Date2?: string;
  Date3?: string;
  Date4?: string;
  Decimal1?: number;
  Decimal2?: number;
  Int1?: number;
  Int2?: number;
  Lookup1Id?: number;
  Lookup2Id?: number;
  Lookup3Id?: number;
  Lookup4Id?: number;
  Lookup5Id?: number;
  Lookup6Id?: number;
  Lookup7Id?: number;
  Lookup8Id?: number;
  Text1?: string;
  Text2?: string;
  Text3?: string;
  Text4?: string;
  Image?: string;

  AgentId?: number;
  TaxIdentificationNumber?: string;
  JobId?: number;
  BankAccountNumber?: number;
  Users?: TRelationUser[];
  Attachments?: TAttachment[];
}

export interface Relation extends RelationForSave<RelationUser, RelationAttachment> {
  DefinitionId?: number;
  ImageId?: string;
  IsActive?: boolean;
  CreatedAt?: string;
  CreatedById?: number | string;
  ModifiedAt?: string;
  ModifiedById?: number | string;
}

const _select = ['', '2', '3'].map(pf => 'Name' + pf);
let _settings: SettingsForClient;
let _definitions: DefinitionsForClient;
let _cache: { [defId: number]: EntityDescriptor } = {};
let _definitionIds: number[];

export function metadata_Relation(wss: WorkspaceService, trx: TranslateService, definitionId: number): EntityDescriptor {
  const ws = wss.currentTenant;
  // Some global values affect the result, we check here if they have changed, otherwise we return the cached result
  if (ws.settings !== _settings || ws.definitions !== _definitions) {
    _settings = ws.settings;
    _definitions = ws.definitions;
    _definitionIds = null;

    // clear the cache
    _cache = {};
  }

  const key = definitionId || '-'; // undefined
  if (!_cache[key]) {
    if (!_definitionIds) {
      _definitionIds = Object.keys(ws.definitions.Relations).map(e => +e);
    }

    const entityDesc: EntityDescriptor = {
      collection: 'Relation',
      definitionId,
      definitionIds: _definitionIds,
      titleSingular: () => ws.getMultilingualValueImmediate(ws.definitions.Relations[definitionId], 'TitleSingular') || trx.instant('Relation'),
      titlePlural: () => ws.getMultilingualValueImmediate(ws.definitions.Relations[definitionId], 'TitlePlural') || trx.instant('Relations'),
      select: _select,
      apiEndpoint: !!definitionId ? `relations/${definitionId}` : 'relations',
      masterScreenUrl: !!definitionId ? `relations/${definitionId}` : 'relations',
      orderby: () => ws.isSecondaryLanguage ? [_select[1], _select[0]] : ws.isTernaryLanguage ? [_select[2], _select[0]] : [_select[0]],
      inactiveFilter: 'IsActive eq true',
      format: (item: EntityWithKey) => ws.getMultilingualValueImmediate(item, _select[0]),
      properties: {
        Id: { datatype: 'integral', control: 'number', label: () => trx.instant('Id'), minDecimalPlaces: 0, maxDecimalPlaces: 0 },
        DefinitionId: { datatype: 'integral', control: 'number', label: () => `${trx.instant('Definition')} (${trx.instant('Id')})`, minDecimalPlaces: 0, maxDecimalPlaces: 0 },
        Definition: { datatype: 'entity', label: () => trx.instant('Definition'), control: 'RelationDefinition', foreignKeyName: 'DefinitionId' },
        Name: { datatype: 'string', control: 'text', label: () => trx.instant('Name') + ws.primaryPostfix },
        Name2: { datatype: 'string', control: 'text', label: () => trx.instant('Name') + ws.secondaryPostfix },
        Name3: { datatype: 'string', control: 'text', label: () => trx.instant('Name') + ws.ternaryPostfix },
        Code: { datatype: 'string', control: 'text', label: () => trx.instant('Code') },
        CurrencyId: { datatype: 'string', control: 'text', label: () => `${trx.instant('Entity_Currency')} (${trx.instant('Id')})` },
        Currency: { datatype: 'entity', label: () => trx.instant('Entity_Currency'), control: 'Currency', foreignKeyName: 'CurrencyId' },
        CenterId: { datatype: 'integral', control: 'number', label: () => `${trx.instant('Entity_Center')} (${trx.instant('Id')})`, minDecimalPlaces: 0, maxDecimalPlaces: 0 },
        Center: { datatype: 'entity', label: () => trx.instant('Entity_Center'), control: 'Center', foreignKeyName: 'CenterId' },
        Description: { datatype: 'string', control: 'text', label: () => trx.instant('Description') + ws.primaryPostfix },
        Description2: { datatype: 'string', control: 'text', label: () => trx.instant('Description') + ws.secondaryPostfix },
        Description3: { datatype: 'string', control: 'text', label: () => trx.instant('Description') + ws.ternaryPostfix },
        LocationJson: { datatype: 'string', control: 'text', label: () => trx.instant('Entity_LocationJson') },
        FromDate: { datatype: 'date', control: 'date', label: () => trx.instant('Entity_FromDate') },
        ToDate: { datatype: 'date', control: 'date', label: () => trx.instant('Entity_ToDate') },

        DateOfBirth: { datatype: 'date', control: 'date', label: () => trx.instant('Relation_DateOfBirth') },
        ContactEmail: { datatype: 'string', control: 'text', label: () => trx.instant('Entity_ContactEmail') },
        ContactMobile: { datatype: 'string', control: 'text', label: () => trx.instant('Entity_ContactMobile') },
        NormalizedContactMobile: { datatype: 'string', control: 'text', label: () => trx.instant('Entity_NormalizedContactMobile') },
        ContactAddress: { datatype: 'string', control: 'text', label: () => trx.instant('Entity_ContactAddress') },
        Date1: { datatype: 'date', control: 'date', label: () => trx.instant('Entity_Date1') },
        Date2: { datatype: 'date', control: 'date', label: () => trx.instant('Entity_Date2') },
        Date3: { datatype: 'date', control: 'date', label: () => trx.instant('Entity_Date3') },
        Date4: { datatype: 'date', control: 'date', label: () => trx.instant('Entity_Date4') },
        Decimal1: { datatype: 'decimal', control: 'number', label: () => trx.instant('Entity_Decimal1'), minDecimalPlaces: 0, maxDecimalPlaces: 4 },
        Decimal2: { datatype: 'decimal', control: 'number', label: () => trx.instant('Entity_Decimal2'), minDecimalPlaces: 0, maxDecimalPlaces: 4 },
        Int1: { datatype: 'integral', control: 'number', label: () => trx.instant('Entity_Int1'), minDecimalPlaces: 0, maxDecimalPlaces: 0 },
        Int2: { datatype: 'integral', control: 'number', label: () => trx.instant('Entity_Int2'), minDecimalPlaces: 0, maxDecimalPlaces: 0 },
        Lookup1Id: { datatype: 'integral', control: 'number', label: () => `${trx.instant('Entity_Lookup1')} (${trx.instant('Id')})`, minDecimalPlaces: 0, maxDecimalPlaces: 0 },
        Lookup1: { datatype: 'entity', label: () => trx.instant('Entity_Lookup1'), control: 'Lookup', foreignKeyName: 'Lookup1Id' },
        Lookup2Id: { datatype: 'integral', control: 'number', label: () => `${trx.instant('Entity_Lookup2')} (${trx.instant('Id')})`, minDecimalPlaces: 0, maxDecimalPlaces: 0 },
        Lookup2: { datatype: 'entity', label: () => trx.instant('Entity_Lookup2'), control: 'Lookup', foreignKeyName: 'Lookup2Id' },
        Lookup3Id: { datatype: 'integral', control: 'number', label: () => `${trx.instant('Entity_Lookup3')} (${trx.instant('Id')})`, minDecimalPlaces: 0, maxDecimalPlaces: 0 },
        Lookup3: { datatype: 'entity', label: () => trx.instant('Entity_Lookup3'), control: 'Lookup', foreignKeyName: 'Lookup3Id' },
        Lookup4Id: { datatype: 'integral', control: 'number', label: () => `${trx.instant('Entity_Lookup4')} (${trx.instant('Id')})`, minDecimalPlaces: 0, maxDecimalPlaces: 0 },
        Lookup4: { datatype: 'entity', label: () => trx.instant('Entity_Lookup4'), control: 'Lookup', foreignKeyName: 'Lookup4Id' },
        Lookup5Id: { datatype: 'integral', control: 'number', label: () => `${trx.instant('Entity_Lookup5')} (${trx.instant('Id')})`, minDecimalPlaces: 0, maxDecimalPlaces: 0 },
        Lookup5: { datatype: 'entity', label: () => trx.instant('Entity_Lookup5'), control: 'Lookup', foreignKeyName: 'Lookup5Id' },
        Lookup6Id: { datatype: 'integral', control: 'number', label: () => `${trx.instant('Entity_Lookup6')} (${trx.instant('Id')})`, minDecimalPlaces: 0, maxDecimalPlaces: 0 },
        Lookup6: { datatype: 'entity', label: () => trx.instant('Entity_Lookup6'), control: 'Lookup', foreignKeyName: 'Lookup6Id' },
        Lookup7Id: { datatype: 'integral', control: 'number', label: () => `${trx.instant('Entity_Lookup7')} (${trx.instant('Id')})`, minDecimalPlaces: 0, maxDecimalPlaces: 0 },
        Lookup7: { datatype: 'entity', label: () => trx.instant('Entity_Lookup7'), control: 'Lookup', foreignKeyName: 'Lookup7Id' },
        Lookup8Id: { datatype: 'integral', control: 'number', label: () => `${trx.instant('Entity_Lookup8')} (${trx.instant('Id')})`, minDecimalPlaces: 0, maxDecimalPlaces: 0 },
        Lookup8: { datatype: 'entity', label: () => trx.instant('Entity_Lookup8'), control: 'Lookup', foreignKeyName: 'Lookup8Id' },
        Text1: { datatype: 'string', control: 'text', label: () => trx.instant('Entity_Text1') },
        Text2: { datatype: 'string', control: 'text', label: () => trx.instant('Entity_Text2') },
        Text3: { datatype: 'string', control: 'text', label: () => trx.instant('Entity_Text3') },
        Text4: { datatype: 'string', control: 'text', label: () => trx.instant('Entity_Text4') },

        // Relation Only

        AgentId: { datatype: 'integral', control: 'number', label: () => `${trx.instant('Relation_Agent')} (${trx.instant('Id')})`, minDecimalPlaces: 0, maxDecimalPlaces: 0 },
        Agent: { datatype: 'entity', label: () => trx.instant('Relation_Agent'), control: 'Agent', foreignKeyName: 'AgentId' },
        TaxIdentificationNumber: { datatype: 'string', control: 'text', label: () => trx.instant('Relation_TaxIdentificationNumber') },
        JobId: { datatype: 'integral', control: 'number', label: () => `${trx.instant('Relation_Job')} (${trx.instant('Id')})`, minDecimalPlaces: 0, maxDecimalPlaces: 0 },
        BankAccountNumber: { datatype: 'string', control: 'text', label: () => trx.instant('Relation_BankAccountNumber') },
        Relation1Id: { datatype: 'integral', control: 'number', label: () => `${trx.instant('Relation_Relation1')} (${trx.instant('Id')})`, minDecimalPlaces: 0, maxDecimalPlaces: 0 },
        Relation1: { datatype: 'entity', label: () => trx.instant('Relation_Relation1'), control: 'Relation', foreignKeyName: 'Relation1Id' },

        // Standard

        IsActive: { datatype: 'boolean', control: 'boolean', label: () => trx.instant('IsActive') },
        CreatedAt: { datatype: 'datetimeoffset', control: 'datetime', label: () => trx.instant('CreatedAt') },
        CreatedBy: { datatype: 'entity', control: 'User', label: () => trx.instant('CreatedBy'), foreignKeyName: 'CreatedById' },
        ModifiedAt: { datatype: 'datetimeoffset', control: 'datetime', label: () => trx.instant('ModifiedAt') },
        ModifiedBy: { datatype: 'entity', control: 'User', label: () => trx.instant('ModifiedBy'), foreignKeyName: 'ModifiedById' }
      }
    };

    if (!ws.settings.SecondaryLanguageId) {
      delete entityDesc.properties.Name2;
      delete entityDesc.properties.Description2;
    }

    if (!ws.settings.TernaryLanguageId) {
      delete entityDesc.properties.Name3;
      delete entityDesc.properties.Description3;
    }

    // Adjust according to definitions
    const definition = _definitions.Relations[definitionId];
    if (!definition) {
      if (!!definitionId) {
        // Programmer mistake
        console.error(`defintionId '${definitionId}' doesn't exist`);
      }
    } else {

      // Definition specific adjustments
      if (definition.State === 'Archived') {
          entityDesc.isArchived = true;
      }

      delete entityDesc.properties.DefinitionId;
      delete entityDesc.properties.Definition;

      // Description, special case
      if (!definition.DescriptionVisibility) {
          delete entityDesc.properties.Description;
          delete entityDesc.properties.Description2;
          delete entityDesc.properties.Description3;
      }

      // Location, special case
      if (!definition.LocationVisibility) {
          delete entityDesc.properties.LocationJson;
      }

      // Simple properties Visibility
      for (const propName of ['TaxIdentificationNumber', 'BankAccountNumber', 'DateOfBirth', 'ContactEmail', 'ContactMobile', 'ContactAddress']) {
          if (!definition[propName + 'Visibility']) {
              delete entityDesc.properties[propName];
          }
      }

      // Special case
      if (!definition.ContactMobileVisibility) {
        delete entityDesc.properties.NormalizedContactMobile;
      }

      // Simple properties Visibility + Label
      for (const propName of ['FromDate', 'ToDate', 'Decimal1', 'Decimal2', 'Int1', 'Int2', 'Text1', 'Text2', 'Text3', 'Text4', 'Date1', 'Date2', 'Date3', 'Date4', 'Identifier']) {
          if (!definition[propName + 'Visibility']) {
              delete entityDesc.properties[propName];
          } else {
              const propDesc = entityDesc.properties[propName] as NavigationPropDescriptor;
              const defaultLabel = propDesc.label;
              propDesc.label = () => ws.getMultilingualValueImmediate(definition, propName + 'Label') || defaultLabel();
          }
      }

      // Navigation properties
      for (const propName of ['Currency', 'Center', 'Agent'/*, 'Job'*/]) {
          if (!definition[propName + 'Visibility']) {
              delete entityDesc.properties[propName];
              delete entityDesc.properties[propName + 'Id'];
          } else {
              const propDesc = entityDesc.properties[propName] as NavigationPropDescriptor;
              const defaultLabel = propDesc.label;
              propDesc.label = () => ws.getMultilingualValueImmediate(definition, propName + 'Label') || defaultLabel();

              const idPropDesc = entityDesc.properties[propName + 'Id'] as NumberPropDescriptor;
              idPropDesc.label = () => `${propDesc.label()} (${trx.instant('Id')})`;
          }
      }

      // Navigation properties with definition Id
      for (const propName of ['1', '2', '3', '4', '5', '6', '7', '8'].map(pf => 'Lookup' + pf).concat(['Relation1'])) {
          if (!definition[propName + 'Visibility']) {
              delete entityDesc.properties[propName];
              delete entityDesc.properties[propName + 'Id'];
          } else {
              const propDesc = entityDesc.properties[propName] as NavigationPropDescriptor;
              propDesc.definitionId = definition[propName + 'DefinitionId'];
              const defaultLabel = propDesc.label;
              propDesc.label = () => ws.getMultilingualValueImmediate(definition, propName + 'Label') || defaultLabel();

              const idPropDesc = entityDesc.properties[propName + 'Id'] as NumberPropDescriptor;
              idPropDesc.label = () => `${propDesc.label()} (${trx.instant('Id')})`;
          }
      }
    }

    _cache[key] = entityDesc;
  }

  return _cache[key];
}

File: Tellma/Entities/Relation.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace Tellma.Entities
{
    [StrongEntity]
    [EntityDisplay(Singular = "Relation", Plural = "Relations")]
    public class RelationForSaveBase<TRelationUser> : EntityWithKey<int>, ILocationEntityForSave, IEntityWithImageForSave
    {
        [MultilingualDisplay(Name = "Name", Language = Language.Primary)]
        [Required]
        [StringLength(255)]
        [AlwaysAccessible]
        public string Name { get; set; }

        [MultilingualDisplay(Name = "Name", Language = Language.Secondary)]
        [StringLength(255)]
        [AlwaysAccessible]
        public string Name2 { get; set; }

        [MultilingualDisplay(Name = "Name", Language = Language.Ternary)]
        [StringLength(255)]
        [AlwaysAccessible]
        public string Name3 { get; set; }

        [Display(Name = "Code")]
        [StringLength(50)]
        [AlwaysAccessible]
        public string Code { get; set; }

        #region Common with Resource

        [Display(Name = "Entity_Currency")]
        [StringLength(3)]
        public string CurrencyId { get; set; }

        [Display(Name = "Entity_Center")]
        public int? CenterId { get; set; }

        [MultilingualDisplay(Name = "Description", Language = Language.Primary)]
        [StringLength(2048)]
        [AlwaysAccessible]
        public string Description { get; set; }

        [MultilingualDisplay(Name = "Description", Language = Language.Secondary)]
        [StringLength(2048)]
        [AlwaysAccessible]
        public string Description2 { get; set; }

        [MultilingualDisplay(Name = "Description", Language = Language.Ternary)]
        [StringLength(2048)]
        [AlwaysAccessible]
        public string Description3 { get; set; }

        [Display(Name = "Entity_LocationJson")]
        public string LocationJson { get; set; }

        // Auto computed from the GeoJSON property, not visible to clients
        public byte[] LocationWkb { get; set; }

        [Display(Name = "Entity_FromDate")]
        public DateTime? FromDate { get; set; }

        [Display(Name = "Entity_ToDate")]
        public DateTime? ToDate { get; set; }

        [Display(Name = "Relation_DateOfBirth")]
        public DateTime? DateOfBirth { get; set; }

        [Display(Name = "Entity_ContactEmail")]
        [EmailAddress]
        [StringLength(255)]
        public string ContactEmail { get; set; }

        [Display(Name = "Entity_ContactMobile")]
        [Phone]
        [StringLength(50)]
        public string ContactMobile { get; set; }

        [Display(Name = "Entity_NormalizedContactMobile")]
        public string NormalizedContactMobile { get; set; }

        [Display(Name = "Entity_ContactAddress")]
        [StringLength(255)]
        public string ContactAddress { get; set; }

        [Display(Name = "Entity_Date1")]
        public DateTime? Date1 { get; set; }

        [Display(Name = "Entity_Date2")]
        public DateTime? Date2 { get; set; }

        [Display(Name = "Entity_Date3")]
        public DateTime? Date3 { get; set; }

        [Display(Name = "Entity_Date4")]
        public DateTime? Date4 { get; set; }

        [Display(Name = "Entity_Decimal1")]
        public decimal? Decimal1 { get; set; }

        [Display(Name = "Entity_Decimal2")]
        public decimal? Decimal2 { get; set; }

        [Display(Name = "Entity_Int1")]
        public int? Int1 { get; set; }

        [Display(Name = "Entity_Int2")]
        public int? Int2 { get; set; }

        [Display(Name = "Entity_Lookup1")]
        public int? Lookup1Id { get; set; }

        [Display(Name = "Entity_Lookup2")]
        public int? Lookup2Id { get; set; }

        [Display(Name = "Entity_Lookup3")]
        public int? Lookup3Id { get; set; }

        [Display(Name = "Entity_Lookup4")]
        public int? Lookup4Id { get; set; }

        [Display(Name = "Entity_Lookup5")]
        public int? Lookup5Id { get; set; }

        [Display(Name = "Entity_Lookup6")]
        public int? Lookup6Id { get; set; }

        [Display(Name = "Entity_Lookup7")]
        public int? Lookup7Id { get; set; }

        [Display(Name = "Entity_Lookup8")]
        public int? Lookup8Id { get; set; }

        [Display(Name = "Entity_Text1")]
        [StringLength(50)]
        public string Text1 { get; set; }

        [Display(Name = "Entity_Text2")]
        [StringLength(50)]
        public string Text2 { get; set; }

        [Display(Name = "Entity_Text3")]
        [StringLength(50)]
        public string Text3 { get; set; }

        [Display(Name = "Entity_Text4")]
        [StringLength(50)]
        public string Text4 { get; set; }

        [NotMapped]
        [Display(Name = "Image")]
        public byte[] Image { get; set; }

        #endregion

        #region Relation Only

        [Display(Name = "Relation_Agent")]
        public int? AgentId { get; set; }

        [Display(Name = "Relation_TaxIdentificationNumber")]
        [StringLength(18)]
        public string TaxIdentificationNumber { get; set; }

        [Display(Name = "Relation_Job")]
        public int? JobId { get; set; }

        [Display(Name = "Relation_BankAccountNumber")]
        [StringLength(34)]
        public string BankAccountNumber { get; set; }

        [NotMapped]
        public int? Relation1Index { get; set; }

        [Display(Name = "Relation_Relation1")]
        [SelfReferencing(nameof(Relation1Index))]
        public int? Relation1Id { get; set; }

        [Display(Name = "Relation_Users")]
        [ForeignKey(nameof(RelationUser.RelationId))]
        public List<TRelationUser> Users { get; set; }

        #endregion
    }

    public class RelationForSave : RelationForSaveBase<RelationUserForSave>
    {
    }

    public class Relation : RelationForSaveBase<RelationUser>, ILocationEntity, IEntityWithImage
    {
        #region Common with Resource

        [Display(Name = "Definition")]
        public int? DefinitionId { get; set; }

        public string ImageId { get; set; }

        [Display(Name = "IsActive")]
        [AlwaysAccessible]
        public bool? IsActive { get; set; }

        [Display(Name = "CreatedAt")]
        public DateTimeOffset? CreatedAt { get; set; }

        [Display(Name = "CreatedBy")]
        public int? CreatedById { get; set; }

        [Display(Name = "ModifiedAt")]
        public DateTimeOffset? ModifiedAt { get; set; }

        [Display(Name = "ModifiedBy")]
        public int? ModifiedById { get; set; }

        // For Query

        [Display(Name = "Definition")]
        [ForeignKey(nameof(DefinitionId))]
        public RelationDefinition Definition { get; set; }

        public Geography Location { get; set; }

        [Display(Name = "Entity_Center")]
        [ForeignKey(nameof(CenterId))]
        public Center Center { get; set; }

        [Display(Name = "CreatedBy")]
        [ForeignKey(nameof(CreatedById))]
        public User CreatedBy { get; set; }

        [Display(Name = "ModifiedBy")]
        [ForeignKey(nameof(ModifiedById))]
        public User ModifiedBy { get; set; }

        [Display(Name = "Entity_Currency")]
        [ForeignKey(nameof(CurrencyId))]
        public Currency Currency { get; set; }

        [Display(Name = "Entity_Lookup1")]
        [ForeignKey(nameof(Lookup1Id))]
        public Lookup Lookup1 { get; set; }

        [Display(Name = "Entity_Lookup2")]
        [ForeignKey(nameof(Lookup2Id))]
        public Lookup Lookup2 { get; set; }

        [Display(Name = "Entity_Lookup3")]
        [ForeignKey(nameof(Lookup3Id))]
        public Lookup Lookup3 { get; set; }

        [Display(Name = "Entity_Lookup4")]
        [ForeignKey(nameof(Lookup4Id))]
        public Lookup Lookup4 { get; set; }

        [Display(Name = "Entity_Lookup5")]
        [ForeignKey(nameof(Lookup5Id))]
        public Lookup Lookup5 { get; set; }

        [Display(Name = "Entity_Lookup6")]
        [ForeignKey(nameof(Lookup6Id))]
        public Lookup Lookup6 { get; set; }

        [Display(Name = "Entity_Lookup7")]
        [ForeignKey(nameof(Lookup7Id))]
        public Lookup Lookup7 { get; set; }

        [Display(Name = "Entity_Lookup8")]
        [ForeignKey(nameof(Lookup8Id))]
        public Lookup Lookup8 { get; set; }

        #endregion

        #region Relation Only

        [Display(Name = "Relation_Agent")]
        [ForeignKey(nameof(AgentId))]
        public Agent Agent { get; set; }

        [Display(Name = "Relation_Relation1")]
        [ForeignKey(nameof(Relation1Id))]
        public Relation Relation1 { get; set; }

        #endregion
    }
}